hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3AFileSystem.java

```diff
@@ -896,45 +941,108 @@ private boolean innerRename(Path source, Path dest)
         keysToDelete.add(new DeleteObjectsRequest.KeyVersion(dstKey));
       }
 
-      Path parentPath = keyToPath(srcKey);
-      RemoteIterator<LocatedFileStatus> iterator = listFilesAndEmptyDirectories(
-          parentPath, true);
-      while (iterator.hasNext()) {
-        LocatedFileStatus status = iterator.next();
-        long length = status.getLen();
-        String key = pathToKey(status.getPath());
-        if (status.isDirectory() && !key.endsWith("/")) {
-          key += "/";
-        }
-        keysToDelete
-            .add(new DeleteObjectsRequest.KeyVersion(key));
-        String newDstKey =
-            dstKey + key.substring(srcKey.length());
-        copyFile(key, newDstKey, length);
-
-        if (hasMetadataStore()) {
-          // with a metadata store, the object entries need to be updated,
-          // including, potentially, the ancestors
-          Path childSrc = keyToQualifiedPath(key);
-          Path childDst = keyToQualifiedPath(newDstKey);
-          if (objectRepresentsDirectory(key, length)) {
-            S3Guard.addMoveDir(metadataStore, srcPaths, dstMetas, childSrc,
-                childDst, username);
+      // A blocking queue that tracks all objects that need to be deleted
+      BlockingQueue<Optional<DeleteObjectsRequest.KeyVersion>> deleteQueue = new ArrayBlockingQueue<>(
+              MAX_ENTRIES_TO_DELETE * 1.5);
+
+      // Used to track if the delete thread was gracefully shutdown
+      boolean deleteFutureComplete = false;
+      FutureTask<Void> deleteFuture = null;
+
+      try {
+        // Launch a thread that will read from the deleteQueue and batch delete any files that have already been copied
+        deleteFuture = new FutureTask<>(new Callable<Void>() {
```

Comment:
If we target java 8 only, this can be a proper lambda-expression :)

Reply:
Changed to lambda expression.